cmake/KokkosConfigCommon.cmake.in

```diff
@@ -203,6 +204,463 @@ int main()
+#
+function(kokkos_defer_set_dependent_library_properties)
+  cmake_parse_arguments(ARG "" "" "LIBRARY" ${ARGN})
+
+  if(ARG_UNPARSED_ARGUMENTS)
+    message(
+      FATAL_ERROR
+        "'kokkos_defer_set_dependent_library_properties' has unrecognized arguments: ${ARG_UNPARSED_ARGUMENTS}"
+    )
+  endif()
+
+  if(NOT ARG_LIBRARY)
+    message(FATAL_ERROR "'kokkos_defer_set_dependent_library_properties' needs a LIBRARY argument")
+  endif()
+
+  cmake_language(
+    EVAL
+    CODE
+    "cmake_language(DEFER DIRECTORY ${CMAKE_SOURCE_DIR} CALL kokkos_set_dependent_library_properties LIBRARY ${ARG_LIBRARY})"
+  )
+
+endfunction()
+
+# this function sets the source and target properties that Kokkos requires
+#
+#       GLOBAL                  --> all targets that link to kokkos in the project
+#       DIRECTORY               --> all targets that link to kokkos in given directories
+#       TARGET                  --> on the given targets (that need compilation and link settings)
+#       LINK_ONLY_TARGET        --> on the given targets (that only need link settings)
+#       Kokkos_FOUND_DIR        --> hint on where Kokkos was installed (important for CMake contexts that don't know this but depend on Kokkos)
+#
+function(kokkos_set_target_properties)
+  cmake_parse_arguments(INP "GLOBAL" "Kokkos_FOUND_DIR" "DIRECTORY;TARGET;LINK_ONLY_TARGET" ${ARGN})
+
+  if(INP_UNPARSED_ARGUMENTS)
+    message(FATAL_ERROR "'kokkos_set_target_properties' has unrecognized arguments: ${INP_UNPARSED_ARGUMENTS}")
+  endif()
+
+  if(KOKKOS_CXX_COMPILER_IS_COMPATIBLE OR @Kokkos_ENABLE_COMPILE_AS_CMAKE_LANGUAGE@)
+    return()
+  endif()
+
+  if(NOT Kokkos_COMPILE_LAUNCHER)
+    # find kokkos_launch_compiler
+    find_program(
+      Kokkos_COMPILE_LAUNCHER
+      NAMES kokkos_launch_compiler
+      HINTS ${INP_Kokkos_FOUND_DIR};${Kokkos_ROOT}
+      PATHS ${INP_Kokkos_FOUND_DIR};${Kokkos_ROOT}
+      PATH_SUFFIXES bin
+    )
+  endif()
+
+  if(NOT Kokkos_COMPILE_LAUNCHER)
+    message(
+      FATAL_ERROR
+        "Kokkos could not find 'kokkos_launch_compiler'. Please set '-DKokkos_COMPILE_LAUNCHER=/path/to/launcher' or '-DKokkos_ROOT=/path/to/install'"
+    )
+  endif()
+
+  if(NOT Kokkos_NVCC_WRAPPER)
+    # find nvcc_wrapper
+    find_program(
+      Kokkos_NVCC_WRAPPER
+      NAMES nvcc_wrapper
+      HINTS ${INP_Kokkos_FOUND_DIR};${Kokkos_ROOT}
+      PATHS ${INP_Kokkos_FOUND_DIR};${Kokkos_ROOT}
+      PATH_SUFFIXES bin
+    )
+  endif()
+  # fatal if we can't nvcc_wrapper
+  if(NOT Kokkos_NVCC_WRAPPER)
+    message(
+      FATAL_ERROR
+        "Kokkos could not find nvcc_wrapper. Please set '-DKokkos_NVCC_WRAPPER=/path/to/nvcc_wrapper' or '-DKokkos_ROOT=/path/to/install'"
+    )
+  endif()
+
+  set(COMPILER_LAUNCHER
+      ${CMAKE_CXX_COMPILER_LAUNCHER};${Kokkos_COMPILE_LAUNCHER};${Kokkos_NVCC_WRAPPER};${CMAKE_CXX_COMPILER}
+  )
+  set(LINKER_LAUNCHER
+      ${CMAKE_CXX_LINKER_LAUNCHER};${Kokkos_COMPILE_LAUNCHER};${Kokkos_NVCC_WRAPPER};${CMAKE_CXX_COMPILER}
+  )
```

Comment:
I suggest commenting here that the trailing `${CMAKE_CXX_COMPILER}` is an argument to the wrapper.  Otherwise it looks like an accidental duplicate because CMake puts this all in front of its generated `${CMAKE_CXX_COMPILER}` call.